datafusion/physical-plan/src/aggregates/group_values/single_group_by/primitive.rs

```diff
@@ -205,20 +282,360 @@ where
+            let single_block = self.values.back_mut().unwrap();
+            single_block.clear();
+            single_block.shrink_to(count);
+        } else {
+            self.values.clear();
+        }
+
+        // Clear mappings
         self.map.clear();
         self.map.shrink_to(count, |_| 0); // hasher does not matter since the map is cleared
+        self.null_group = None;
+
+        // Clear helping structures
+        self.emitting = false;
+        self.num_groups = 0;
+    }
+
+    fn supports_blocked_groups(&self) -> bool {
+        true
+    }
+
+    fn alter_block_size(&mut self, block_size: Option<usize>) -> Result<()> {
+        // Clear values
+        self.values.clear();
+
+        // Clear mappings
+        self.map.clear();
+        self.null_group = None;
+
+        // Clear helping structures
+        self.emitting = false;
+        self.num_groups = 0;
+
+        // As mentioned above, we ensure the `single block` always exist
+        // in `flat mode`
+        if block_size.is_none() {
+            self.values.push_back(Vec::new());
+        }
+        self.block_size = block_size;
+
+        Ok(())
+    }
+}
+
+impl<T: ArrowPrimitiveType> GroupValuesPrimitive<T>
+where
+    T::Native: HashValue,
+{
+    fn get_or_create_groups<F, O>(
+        &mut self,
+        cols: &[ArrayRef],
+        groups: &mut Vec<usize>,
+        mut before_add_group: F,
+    ) -> Result<()>
+    where
+        F: FnMut(&mut VecDeque<Vec<T::Native>>),
+        O: GroupIndexOperations,
+    {
+        assert_eq!(cols.len(), 1);
+        groups.clear();
+
+        let block_size = self.block_size.unwrap_or_default();
+        for v in cols[0].as_primitive::<T>() {
+            let group_index = match v {
+                None => *self.null_group.get_or_insert_with(|| {
+                    // Actions before add new group like checking if room is enough
+                    before_add_group(&mut self.values);
+
+                    // Get block infos and update block,
+                    // we need `current block` and `next offset in block`
+                    let current_block = self.values.back_mut().unwrap();
+                    current_block.push(Default::default());
+
+                    // Compute group index
+                    let group_index = self.num_groups;
+                    self.num_groups += 1;
+
+                    // Get group index and finish actions needed it
+                    group_index
+                }),
+                Some(key) => {
+                    let state = &self.random_state;
+                    let hash = key.hash(state);
+                    let insert = self.map.entry(
+                        hash,
+                        |g| unsafe {
+                            let block_id = O::get_block_id(g.0, block_size);
+                            let block_offset = O::get_block_offset(g.0, block_size);
+                            self.values[block_id].get_unchecked(block_offset).is_eq(key)
```

Comment:
Should we use `get_unchecked` here as well?

I wonder if it is worth it to store a reference to the block in the map to save the first lookup in `self.values`?

Reply:
> Should we use get_unchecked here as well for self.values?
I tried it before, seems `Vec::get_unchecked` no obvious improvement in my local.
I am trying it again now.

> I wonder if it is worth it to store a reference to the block in the map to save the first lookup in self.values?
I think with `Vec<Vec<T>>`, the twice lookup is acceptable, `index lookup` seems really efficient.
As I see, the reason why `index lookup` of `VecDeque` expansive, is due to its more complex index computation, and instead only simple pointer add is needed in `Vec`.

Reply:
Many many codes in one `index lookup` of `VecDeque`... `Vec` is really friendly structure...

```rust
  fn to_physical_idx(&self, idx: usize) -> usize {
      self.wrap_add(self.head, idx)
  }

  fn wrap_add(&self, idx: usize, addend: usize) -> usize {
      wrap_index(idx.wrapping_add(addend), self.capacity())
  }

fn wrap_index(logical_index: usize, capacity: usize) -> usize {
    debug_assert!(
        (logical_index == 0 && capacity == 0)
            || logical_index < capacity
            || (logical_index - capacity) < capacity
    );
    if logical_index >= capacity { logical_index - capacity } else { logical_index }
}
```